### PR TITLE
DB-6431: Fix imported rows count and bad records count for IMPORT_DATA and HFILE_BULK_LOAD

### PR DESCRIPTION
**Pull request by wb14123**
This is the first patch to fix one of the bugs in DB-6431.

`AbstractPipelineWriter` already write the rows count to `operationContext` based on `writeStats`. But there are some problems:

1. The `writeStats` returned from `BulkWriteAction` is not correct.
2. `SparkOperationContext` record the written rows to `pipelineRowsWritten`, which seems to be not used in anywhere.
3. Because of 2, `AbstractPipelineWriter` didn't write the row count into `rowsWritten`. Thus some pipeline writer re-add these counters. And when they add these counters, the data is still in buffer which may fail in later writes.

This patch fix these issues.

### CODE REVIEW

File: pipeline_api/src/main/java/com/splicemachine/pipeline/client/BulkWriteAction.java [hunk 1]
```diff
@@ -322,7 +322,9 @@ else if(ctx.attemptCount>100 && ctx.attemptCount%50==0){
                                     ctx.refreshCache=ctx.refreshCache || isFailure;
                                     ctx.sleep=true; //always sleep
                                 }
-                                writtenCounter.add(currentBulkWrite.getSize() - writes.size());
+                                writtenCounter.add(currentBulkWrite.getSize() -
+                                        bulkWriteResult.getNotRunRows().size() -
+                                        bulkWriteResult.getFailedRows().size());
                                 partialRetriedRows.add(writes.size());
```
Comment (jyuanca) [inline pipeline_api/src/main/java/com/splicemachine/pipeline/client/BulkWriteAction.java]: Good catch!

**Comment by wb14123** [comment]
I've merged PR #1632 to here and add another commit to fix HFILE_BULK_LOAD with sampling. Please review the appended two commits. Thanks!

**Comment by wb14123** [comment]
I've add some more changes: check `SparkOperationContext.op` is null before reset the accumulators and rewrite the last commit's description

File: splice_machine/src/test/test-data/rows_count.csv [hunk 2]
```diff
@@ -0,0 +1,3 @@
+"abc","efg"
+"abc","def"
+"abd",null
```
Comment (jyuanca) [inline splice_machine/src/test/test-data/rows_count.csv]: can this small dataset reproduce the error count discrepancies for HFile import with sampling?

Comment (wb14123) [reply splice_machine/src/test/test-data/rows_count.csv]: Yes. With sampling the null row will be count twice.

**Comment by wb14123** [comment]
@dgomezferro Currently `BadRecordsRecorder` only records the number of bad records so I don't see an easy way to reset the bad record files. May be we can fix this in a future issue?

**Comment by jyuanca** [comment]
@dgomezferro Do you have any thoughts about Bin's response?

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkOperationContext.java [hunk 3]
```diff
@@ -221,6 +220,34 @@ public void prepare(){
     @Override
     public void reset(){
 //        impl.resetContextManager();
+        BadRecordsRecorder oldRecorder = this.badRecordsAccumulator.value();
+        BadRecordsRecorder badRecordsRecorder = new BadRecordsRecorder(
+                oldRecorder.getStatusDirectory(),
+                importFileName,
+                badRecordThreshold);
+        this.badRecordsAccumulator=SpliceSpark.getContext().accumulable(badRecordsRecorder,
```
Comment (dgomezferro) [inline hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkOperationContext.java]: This should only be initialized if permissive is true

**Pull request merged**